src/App.tsx

```diff
@@ -1,24 +1,178 @@
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
 
-const USER_ID = 0;
+const USER_ID = 7033;
+
+function getVisibleTodos(todos: Todo[], currentFilter: string) {
+  let visibleTodos = [...todos];
+
+  if (currentFilter !== Filter.All) {
+    visibleTodos = visibleTodos.filter(({ completed }) => {
+      switch (currentFilter) {
+        case Filter.Active:
+          return !completed;
+        case Filter.Completed:
+          return completed;
+        default:
+          return true;
+      }
+    });
+  }
+
+  return visibleTodos;
+}
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterBy, setFilterBy] = useState<Filter>(Filter.All);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [title, setTitle] = useState('');
+  const [loadingTodos, setLoadingTodos] = useState([0]);
+
+  function clearError() {
+    setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  }
+
+  const loadTodos = async () => {
+    try {
+      const todosFromServer = await getTodos(USER_ID);
+
+      setTodos(todosFromServer);
+    } catch (error) {
+      setErrorMessage((error as Error).message);
+
+      clearError();
+    }
+  };
+
+  const visibleTodos = getVisibleTodos(todos, filterBy);
+
+  const addNewTodo = async () => {
+    try {
+      if (title) {
+        const newTodo = {
+          id: 0,
+          userId: USER_ID,
+          title,
+          completed: false,
+        };
+
+        setTempTodo(newTodo);
+        setLoadingTodos(prev => [...prev, newTodo.id]);
+        const addedTodo = await addTodo(USER_ID, newTodo);
+
+        setTodos((current) => [...current, addedTodo]);
+      } else {
+        setErrorMessage('Title can\'t be empty');
+
+        clearError();
+      }
+    } catch {
+      setErrorMessage('Unable to add a todo');
+
+      clearError();
+    } finally {
+      setTitle('');
+      setTempTodo(null);
+      setLoadingTodos([0]);
+    }
+  };
+
+  const removeTodoFromServer = async (todoId: number) => {
+    setLoadingTodos(prev => [...prev, todoId]);
+
+    try {
+      setLoadingTodos(prev => [...prev, todoId]);
+      await deleteTodo(todoId);
+    } catch {
+      setErrorMessage('Unable to delete a todo');
+
+      clearError();
+    } finally {
+      setLoadingTodos([0]);
+    }
+  };
+
+  const removeCompletedTodos = async () => {
+    const completedTodoIds = visibleTodos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    try {
+      setLoadingTodos(prev => [...prev, ...completedTodoIds]);
+      await Promise.all(
+        completedTodoIds.map(id => deleteTodo(id)),
+      );
+    } catch {
+      setErrorMessage('Unable to delete completed todos');
+
+      clearError();
+    } finally {
+      setLoadingTodos([0]);
+    }
+  };
+
+  useEffect(() => {
+    loadTodos();
+  }, [todos]);
```

Comment:
<img width="1375" alt="Screenshot 2023-04-14 at 17 31 58" src="https://user-images.githubusercontent.com/53228259/232073766-3ff8f6d6-4392-49df-8d01-b6de853d00a2.png">
you have a cyclical dependency here due to todos

Reply:
Oops!